Swap order of levels and locations in interpolator result array

twsearle opened the pull request:
## Description

OOPs is increasing efficiency by swapping the order of levels and locations in the generation of geovals. This change is a companion change to adapt to the swap.

## Dependencies

build-group=https://github.com/JCSDA-internal/ufo/pull/3726
build-group=https://github.com/JCSDA-internal/oops/pull/2916

## Impact

Minor memory performance improvements expected from the parent change.

## Issues Resolved

Fixes #141 

## Checklist

- [x] I have updated the unit tests to cover the change
- [x] New functions are documented briefly via Doxygen comments in the code
- [x] I have linted my code using cpplint
- [x] I have run the unit tests
- [x] I have run [mo-bundle](https://cylchub/services/cylc-review/taskjobs/toby.searle/?suite=mobb-oj148) to check integration with the rest of JEDI and run the unit tests under all environments

## Sith testing
 - [ ] [exab_gulf configuration](https://cylchub/services/cylc-review/taskjobs/toby.searle/?&suite=sith_oj148%2Fexab_gulf)
 - [ ] [exab_amm7 configuration](https://cylchub/services/cylc-review/taskjobs/toby.searle/?&suite=sith_oj148%2Fexab_amm7)
 - [ ] [exab_amm15 configuration](https://cylchub/services/cylc-review/taskjobs/toby.searle/?&suite=sith_oj148%2Fexab_amm15)
 - [ ] [exab_ostia configuration](https://cylchub/services/cylc-review/taskjobs/toby.searle/?&suite=sith_oj148%2Fexab_ostia)
 - [ ] [exab_ocnd configuration](https://cylchub/services/cylc-review/taskjobs/toby.searle/?&suite=sith_oj148%2Fexab_ocnd)
 - [ ] [exab_gl_ocn configuration](https://cylchub/services/cylc-review/taskjobs/toby.searle/?&suite=sith_oj148%2Fexab_gl_ocn)

twsearle committed:
Swap order of levels and locations in interpolator

twsearle committed:
Use surface level geometry mask only

twsearle commented:
@frld I believe this change has highlighted a bug with 3DVar and your "gmask" code. This seems to be attempting to make a field 10 levels out of a variable with 1 level on develop. I tried to fix this by ensuring that the geometry mask had the full 10 levels but that caused issues in BUMP.

This led me to the conclusion that SABER/BUMP isn't liking multi-level masks. Instead I set the gmask variable so that it always has a single level. This removes the runtime errors, however the tests now fail as the cost functions no longer match:
```
Test: 'CostJb   : Nonlinear Jb = 0.0000000000000000e+00'
Ref:  'CostJo   : Nonlinear Jo(Sea Ice) = 1.4687500000000000e+00, nobs = 47, Jo/n = 3.1250000000000000e-02, err = 2.0000000000000000e+00'
```

If I run out of time to debug this and you aren't able to get back to me by the end of week, I will force the test ref to match what is produced to get past this error. Otherwise we will be holding up a larger change accross the whole collaboration because our current code reads data off the end of an array. Better to have running tests that don't do what you want than tests that can't run at all without seg faulting.

twsearle commented:
@frld no worries I think I figured it out! It was an issue with an upstream branch being out of date I think. I would still appreciate a review if possible but if not I will push ahead with other reviewers.

ctgh commented on src/orca-jedi/geometry/Geometry.h:
```diff
@@ -75,6 +75,7 @@ class Geometry : public util::Printable {
   std::shared_ptr<eckit::Timer> timer() const {return eckit_timer_;}
   void log_status() const;
   void set_gmask(atlas::Field &) const;
+  const size_t n_levels() const {return n_levels_;}
```
I think it's best to remove the first `const` because you are returning by value.

twsearle replied:
removed completely in latest commit as unused.

ctgh commented on src/orca-jedi/interpolator/Interpolator.cc:
```diff
@@ -2,134 +2,119 @@
-                       << "atlasObsFuncSpace ... done" << std::endl;
-     return atlas::functionspace::PointCloud(std::move(points));
+    const std::vector<double>& lats, const std::vector<double>& lons) {
+  size_t nlocs = lats.size();
```
This can be `const`.

twsearle replied:
done

ctgh commented on src/orca-jedi/interpolator/Interpolator.cc:
```diff
@@ -138,22 +123,19 @@ void Interpolator::apply(const oops::Variables& vars, const State& state,
-    std::vector<double>::iterator& iter) const {
+template <class T>
+void Interpolator::executeInterpolation(
+    const std::string& gv_varname, size_t var_size, const State& state,
```
I believe `var_size` can be `const`. (Same below)

twsearle replied:
these changes have all come in because I stupidly hit "reformat doc" in vscode... I think annoyingly its actually broken Cpplint compliance. Will have to go through.

twsearle replied:
made const.

ctgh commented on src/orca-jedi/interpolator/Interpolator.cc:
```diff
@@ -138,22 +123,19 @@ void Interpolator::apply(const oops::Variables& vars, const State& state,
-      if ( mask[iloc] ) {
+  for (std::size_t iloc = 0; iloc < nlocs_; iloc++) {
+    for (std::size_t klev = 0; klev < var_size; ++klev) {
+      if (mask[iloc]) {
```
You are testing this condition `var_size` times - can you reorder the way this is done to save time?

Similarly, is it necessary to test `is_mv` each time, or would it be more efficient to reorder that?

Also, you could define a missing value and use it instead of the repeated calls to `util::missingValue`.

twsearle replied:
Will do, although repeated calls to util::missingValue are optimised away anyway by compiler I think.

twsearle replied:
Moved if outside of the level loop.

ctgh reviewed (commented):
Looks good - just a few suggestions.

twsearle committed:
Fix unit testing of state and interpolator.

Also includes response to reviews.

twsearle commented:
> Looks good - just a few suggestions.

Thanks very much! I have gone through and addressed your comments (feel free to unresolve conversations above if there are still concerns). I also realised there were some tests I failed to run that needed fixing due to the dimension swap and fix to set_gmask. I will rerun the mo-bundle in the description now.

ctgh reviewed (approved):
Looks good, thanks!

twsearle commented:
This is failing CI because it is a coordinated change and the build-group functionality doesn't work on orca-jedi right now.

fmahebert commented on src/orca-jedi/interpolator/Interpolator.cc:
```diff
@@ -138,78 +123,70 @@ void Interpolator::apply(const oops::Variables& vars, const State& state,
+  for (std::size_t iloc = 0; iloc < nlocs_; iloc++) {
+    if (mask[iloc]) {
+      for (std::size_t klev = 0; klev < var_size; ++klev) {
         if (has_mv && mv(field_view(iloc, klev))) {
```
Naïve question — are these missing values always encoding the 2d mask, or could they have 3d structure? If they're always 2d, there's a chance you could pull the `if` outside the loop over levels.

twsearle replied:
@fmahebert They are guaranteed to have a 3D structure so long as we are using 3D model fields! Its one of our main pain points when engaging with the rest of atlas/JEDI :)

The reason is because our ocean models use missing values to account for bathymetry. You would have to talk to a modeller to get a good answer, but my understanding is that its because bathymetry can affect all the levels unlike orography,  and needs to be masked to model the coastlines.

We use two options from NEMO in our ocean forecasts. The most common is straight-forward, uniform depth levels with masked points for the bathymetry. The more complex option uses hybrid sigma-z levels - this is where the distance between levels is reduced in shallow waters, but also with gradual masking of depth levels as the sea floor rises to the coast.

There are also other options available that aren't currently in use. This image shows them all and maybe easier to understand than the above paragraph.

![image](https://github.com/user-attachments/assets/cfefa79a-4759-446f-b10a-9a952c238bc6)

reference:
https://www.nemo-ocean.eu/doc/node21.html

fmahebert replied:
Makes sense — thanks @twsearle for the instructive response!

fmahebert reviewed (approved):
non-expert review, but looks good

fmahebert commented:
The main OOPS PR was merged, so this can be merged as well.
